homeassistant/components/sensorpush/__init__.py

```diff
@@ -0,0 +1,64 @@
+"""The SensorPush Bluetooth integration."""
+from __future__ import annotations
+
+import logging
+from typing import Optional, Union
+
+from sensorpush_ble import SensorPushBluetoothDeviceData
+
+from homeassistant.components.bluetooth.passive_update_coordinator import (
+    PassiveBluetoothDataUpdate,
+    PassiveBluetoothDataUpdateCoordinator,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.service_info.bluetooth import BluetoothServiceInfo
+
+from .const import DOMAIN
+from .data import sensor_update_to_bluetooth_data_update
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SensorPushDataUpdateCoordinator(
+    PassiveBluetoothDataUpdateCoordinator[Optional[Union[float, int]]]
+):
+    """Coordinator for SensorPush Bluetooth data."""
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up SensorPush BLE device from a config entry."""
+    address = entry.unique_id
+    assert address is not None
+
+    data = SensorPushBluetoothDeviceData()
+
+    @callback
+    def _async_update_data(
+        service_info: BluetoothServiceInfo,
+    ) -> PassiveBluetoothDataUpdate:
+        """Update data from SensorPush Bluetooth."""
+        return sensor_update_to_bluetooth_data_update(data.update(service_info))
+
+    coordinator = hass.data.setdefault(DOMAIN, {})[
+        entry.entry_id
+    ] = SensorPushDataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        update_method=_async_update_data,
+        address=address,
+    )
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(coordinator.async_setup())
```

Comment:
Can this raise? If not, can setup be done automatically on first listener attached?

Reply:
It can't raise.  We could have it setup/teardown automatically when the listener count goes above 1, or to 0

Reply:
I need to change the coordinator back to the track time interval because we are checking async_address_present as well which is also more expensive than I had hoped so I'll need to wrap that in a cache so we only build a dict every 5 minutes or so. When the callback time is reached it just calls in the loop forever right now if the address is still present but hasn't broadcasted anything.  

So I'm going to both of these in a single PR

Reply:
Maybe instead we move this logic into Bluetooth so we can get a callback when the devices goes unavailable. Then we only have one track time interval which is polling the discovered devices every 5 minutes to see if it's gone away, purging the history, and then we can subscribe to unavailable notifications. 

It would solve the overhead problem since history is already indexed on address and also solve history providing stale data.

Reply:
async_step is gone now 4c0e9d066c3b1f423cc13ce49f7f6177834f4775